### PR TITLE
Fix off-by-one in process_pending_consolidations

### PR DESCRIPTION
**Pull request by mkalinin**
Fixes an off-by-one in `process_pending_consolidations`, due to this bug a consolidation processing was delayed by an epoch which compounded with withdrawal sweep resulted in withdrawing the balance to EL instead of consolidating it to the target.

### The fix

The *expected* logic of pending consolidation and deposit processing is as the following:
* Let the `next_epoch` be the epoch the state is being transitioned to
* Process pending consolidation if `source.withdrawable_epoch == next_epoch` to prevent consolidating balance from being withdrawn
* Process pending deposit if `source.withdrawable_epoch == next_epoch + 1` to prevent the balance to be accrued to the consolidating validator (as consolidations are processed after deposits in the epoch processing)

The actual logic uses `current_epoch` instead of the `next_epoch`. Note that with `process_pending_balance_deposits` case this would result in just delaying the processing by an epoch, while with consolidations it would lead to much worse outcome. The corresponding spec tests are added to cover both scenarios.

Huge gratitude to @parithosh and entire devops team for finding this bug on devnet-2. Fantastic catch!! 🙌

### CODE REVIEW

**Review by fradamt — APPROVED** [review]
Lgtm! Your explanation perfectly matches the intended behavior, as far as I can tell from what I have previously noted in the wip annotated spec. 

Nice catch 🐼s :)

File: tests/core/pyspec/eth2spec/test/electra/epoch_processing/test_process_pending_consolidations.py [hunk 1]
```diff
@@ -185,3 +191,153 @@ def test_all_consolidation_cases_together(spec, state):
...
+    expected_source_balance = state_before_consolidation.balances[source_index] - spec.MIN_ACTIVATION_BALANCE
+    assert (
+        state.validators[target_index].withdrawal_credentials[:1]
+        == spec.COMPOUNDING_WITHDRAWAL_PREFIX
+    )
+    assert state.balances[target_index] == 2 * spec.MIN_ACTIVATION_BALANCE
+    assert state.balances[source_index] == expected_source_balance
+    assert state.pending_consolidations == []
+
+    # Pending balance deposit to the target is created
```
[the remark this answers is not in the document]
Comment (mkalinin) [reply tests/core/pyspec/eth2spec/test/electra/epoch_processing/test_process_pending_consolidations.py]: When I was writing this test, finding where the rest of the balance goes took me decent amount of efforts. So, the comment should be really helpful here

File: tests/core/pyspec/eth2spec/test/electra/epoch_processing/test_process_pending_consolidations.py [hunk 2]
```diff
@@ -185,3 +191,158 @@ def test_all_consolidation_cases_together(spec, state):
...
+
+    # Advance to withdrawable_epoch - 1 with full participation
+    target_epoch = state.validators[source_index].withdrawable_epoch - spec.Epoch(1)
+    while spec.get_current_epoch(state) < target_epoch:
+        next_epoch_with_full_participation(spec, state)
+
+    # Obtain state before the call to process_pending_consolidations
+    state_before_consolidation = compute_state_by_epoch_processing_to(spec, state, "process_pending_consolidations")
+
+    yield from run_epoch_processing_with(spec, state, "process_pending_consolidations")
```
Comment (hwwhww) [inline tests/core/pyspec/eth2spec/test/electra/epoch_processing/test_process_pending_consolidations.py:224]: Another solution is to make `run_epoch_processing_with` return values:
```python
def run_epoch_processing_with(spec, state, process_name: str):
    """
    Processes to the next epoch transition, up to and including the sub-transition named ``process_name``
      - pre-state ('pre'), state before calling ``process_name``
      - post-state ('post'), state after calling ``process_name``
    """
    run_epoch_processing_to(spec, state, process_name)
    pre = state.copy() 
    yield 'pre', pre
    getattr(spec, process_name)(state)
    yield 'post', state

    return pre, state
```

and call it with return values in tests:
```python
    pre, post = yield from run_epoch_processing_with(spec, state, "process_pending_consolidations")
```

**Review by hwwhww — APPROVED** [review]
LGTM! Thank you @mkalinin!

Some lines in tests seem a bit repeated, but I also didn't find a way to refactor the fragmented assertions better. 😭

Left a minor suggestion. I can refactor it for tomorrow if you think it makes sense too.

**Pull request merged**